rust/kernel/miscdev.rs

```diff
@@ -87,6 +88,10 @@ impl<T: Sync> FileOpenAdapter for Registration<T> {
     unsafe fn convert(_inode: *mut bindings::inode, file: *mut bindings::file) -> *const Self::Arg {
         // SAFETY: the caller must guarantee that `file` is valid.
         let reg = crate::container_of!(unsafe { (*file).private_data }, Self, mdev);
+        // SAFETY: before calling the `open` callback, `file.private_data` is initialized
+        // by the driver to point to the miscellaneous device, which is guaranteed by
+        // the caller to have been registered through `Registration`, so `file.private_data`
+        // is the `mdev` field of a `Registration` object.
```

Comment:
Is this actually true? All I could find in the kernel docs is this, from the "Overview of the Linux VFS" chapter:

> The open() method is a good place to initialize the “private_data” member in the file structure if you want to point to a device structure

Also I'm a little confused by this line in `open_callback` which happens after `convert` is called https://github.com/Rust-for-Linux/linux/blob/24566f35d0590b3c8b2f41fa0defcb4569f7443f/rust/kernel/file_operations.rs#L105

Reply:
The kernel miscdev subsystem initialises `private_data` prior to calling the driver's `open` (our `open_callback`): https://elixir.bootlin.com/linux/latest/source/drivers/char/misc.c#L136

Our `open_callback` then calls `FileOpener::open` and replaces the contents of `private_data` with the value returned by `FileOpener::open` converted to a pointer via `into_pointer`.

From then on, functions from the `FileOperations` trait will be called by borrowing from this pointer.

Eventually, when all references to the file go away, the value returned by `open` is converted back via `from_pointer` and passed to `FileOperations::release` so that it can be freed.

Reply:
Other subsystems (e.g., `cdev`) store state elsewhere, that's why we have FileOpener.

Reply:
Okay, I see. Thank you very much for the complementary information. Let me correct the comment a little.